src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/ProductController.php

```diff
@@ -211,6 +236,298 @@ public function deleteAction(Request $request, int $productId): Response
         return $this->redirectToRoute('admin_products_v2_index');
     }
 
+    /**
+     * @param Request $request
+     * @param int $productId
+     *
+     * @return Response
+     */
+    public function previewAction(Request $request, int $productId): Response
+    {
+        try {
+            $productUrlProvider = $this->get('prestashop.adapter.shop.url.product_provider');
+            $url = $productUrlProvider->getUrl($productId, '{friendly-url}');
+        } catch (ProductException $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages($e)));
+
+            return $this->redirectToRoute('admin_products_v2_index');
+        }
+
+        return $this->redirect($url);
+    }
+
+    /**
+     * @param Request $request
+     * @param int $productId
+     *
+     * @return Response
+     */
+    public function duplicateAction(Request $request, int $productId): Response
+    {
+        try {
+            $this->getCommandBus()->handle(new DuplicateProductCommand($productId));
+            $this->addFlash(
+                'success',
+                $this->trans('Successful duplication', 'Admin.Notifications.Success')
+            );
+        } catch (ProductException $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages($e)));
+        }
+
+        return $this->redirectToRoute('admin_products_v2_index');
+    }
+
+    /**
+     * Active product ajax action
+     *
+     * @AdminSecurity("is_granted('update', request.get('_legacy_controller'))", redirectRoute="admin_products_v2_index")
+     * @DemoRestricted(redirectRoute="admin_products_v2_index")
+     *
+     * @param Request $request
+     *
+     * @return JsonResponse
+     */
+    public function activateAjaxAction(Request $request): JsonResponse
+    {
+        /** @todo this is temprorary to test errors */
+        $response = [
+            'success' => false,
+            'message' => $this->trans('Test Error.', 'Admin.Notifications.Success')
```

Comment:
What is this message? Could you share the context with me, please? It doesn't look like a success message either 😇